### PR TITLE
Improve logging for agent upgrades.

### PR DESCRIPTION
**Pull request by cmacknz**
Improve our logging around agent upgrades, following the suggestions from https://github.com/elastic/elastic-agent/issues/1253.

The biggest issue I noticed was that we log action failures at the debug label by default, which includes the upgrade action. Upgrade action failures are infrequent and important enough that I switched to always logging them at error level. I think this will solve most of the problems, but I also followed through and added some additional logging in interesting places.

- Closes https://github.com/elastic/elastic-agent/issues/1253

Sample logs from a standalone upgrade, which always fails attempting to execute the upgrade watcher for reasons that aren't clear to me yet: `"fork/exec elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-82ad1e/elastic-agent: no such file or directory"`

```json
{"log.level":"info","@timestamp":"2022-09-25T14:09:59.286-0400","log.origin":{"file.name":"upgrade/cleanup.go","file.line":22},"message":"Cleaning up non-matching downloaded versions","version":"8.6.0","downloads.path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-82ad1e/downloads","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:09:59.323-0400","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2022-09-25T14:09:59-04:00 - message: Application: [7a5f506a-2836-40f4-91eb-70c20bb755b4]: State changed to UPDATING: Update to version '8.4.2' started - type: 'STATE' - sub_type: 'UPDATING'","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:09:59.323-0400","log.origin":{"file.name":"upgrade/step_download.go","file.line":43},"message":"Downloading upgrade artifact","version":"8.4.2","source_uri":"https://artifacts.elastic.co/downloads/","drop_path":"","target_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-82ad1e/downloads","install_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-82ad1e/install","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:10:18.636-0400","log.origin":{"file.name":"http/downloader.go","file.line":307},"message":"download from https://artifacts.elastic.co/downloads/beats/elastic-agent/elastic-agent-8.4.2-darwin-aarch64.tar.gz completed in 18 seconds @ 14.53MBps","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:10:18.645-0400","log.origin":{"file.name":"http/downloader.go","file.line":307},"message":"download from https://artifacts.elastic.co/downloads/beats/elastic-agent/elastic-agent-8.4.2-darwin-aarch64.tar.gz.sha512 completed in Less than a second @ +InfYBps","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:10:22.803-0400","log.origin":{"file.name":"upgrade/step_unpack.go","file.line":39},"message":"Unpacked upgrade artifact","version":"8.4.2","file.path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-82ad1e/downloads/elastic-agent-8.4.2-darwin-aarch64.tar.gz","hash":"d3eb3e","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:10:22.803-0400","log.origin":{"file.name":"upgrade/upgrade.go","file.line":303},"message":"Copying action store","new_home_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-d3eb3e","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:10:22.803-0400","log.origin":{"file.name":"upgrade/step_relink.go","file.line":35},"message":"Changing symlink","symlink_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/elastic-agent","new_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-d3eb3e/elastic-agent","prev_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/elastic-agent.prev","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:10:22.825-0400","log.origin":{"file.name":"upgrade/step_mark.go","file.line":116},"message":"Writing upgrade marker file","file.path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/.update-marker","hash":"d3eb3e","prev_hash":"82ad1e","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:10:22.825-0400","log.origin":{"file.name":"upgrade/step_mark.go","file.line":131},"message":"Updating active commit","file.path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/.elastic-agent.active.commit","hash":"d3eb3e","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2022-09-25T14:10:22.827-0400","log.origin":{"file.name":"upgrade/upgrade.go","file.line":189},"message":"Rolling back: starting watcher failed","error":{"message":"fork/exec /Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-82ad1e/elastic-agent: no such file or directory"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-25T14:10:22.832-0400","log.origin":{"file.name":"upgrade/step_relink.go","file.line":35},"message":"Changing symlink","symlink_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/elastic-agent","new_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/data/elastic-agent-82ad1e/elastic-agent","prev_path":"/Users/cmackenzie/Downloads/elastic-agent-8.5.0-SNAPSHOT-darwin-aarch64/elastic-agent.prev","ecs.version":"1.6.0"}
```

### CODE REVIEW

File: internal/pkg/agent/application/upgrade/cleanup.go [hunk 1]
```diff
@@ -13,11 +13,15 @@ import (
...
+	"github.com/elastic/elastic-agent/pkg/core/logger"
 )
 
-// preUpgradeCleanup will remove files that do not have the passed version number from the downloads directory.
-func preUpgradeCleanup(version string) error {
-	files, err := os.ReadDir(paths.Downloads())
+// cleanNonMatchingVersionsFromDownloads will remove files that do not have the passed version number from the downloads directory.
+func cleanNonMatchingVersionsFromDownloads(log *logger.Logger, version string) error {
+	downloadsPath := paths.Downloads()
+	log.Infow("Cleaning up non-matching downloaded versions", "version", version, "downloads.path", downloadsPath)
```
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/cleanup.go:22]: we were trying to reduce amount of info level logs, consider debug

File: internal/pkg/agent/application/upgrade/step_mark.go [hunk 2]
```diff
@@ -112,20 +113,22 @@ func (u *Upgrader) markUpgrade(_ context.Context, hash string, action Action) er
 	}
 
 	markerPath := markerFilePath()
+	log.Infow("Writing upgrade marker file", "file.path", markerPath, "hash", marker.Hash, "prev_hash", prevHash)
```
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/step_mark.go:116]: consider debug

File: internal/pkg/agent/application/upgrade/step_unpack.go [hunk 3]
```diff
@@ -30,18 +31,21 @@ func (u *Upgrader) unpack(ctx context.Context, version, archivePath string) (str
...
 	if runtime.GOOS == "windows" {
-		hash, err = unzip(version, archivePath)
+		hash, err = unzip(u.log, version, archivePath)
 	} else {
-		hash, err = untar(version, archivePath)
+		hash, err = untar(u.log, version, archivePath)
 	}
+
 	if err != nil {
+		u.log.Infow("Failed to unpack upgrade artifact", "error.message", err, "version", version, "file.path", archivePath, "hash", hash)
```
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/step_unpack.go:40]: Should be an error

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 4]
```diff
@@ -169,39 +170,47 @@ func (u *Upgrader) Upgrade(ctx context.Context, a Action, reexecNow bool) (_ ree
...
+	trimmedNewHash := release.TrimCommit(newHash)
+	cb := shutdownCallback(u.log, paths.Home(), release.Version(), a.Version(), trimmedNewHash)
 	if reexecNow {
+		u.log.Infow("Removing downloads directory", "file.path", paths.Downloads(), "rexec", reexecNow)
 		err = os.RemoveAll(paths.Downloads())
 		if err != nil {
-			u.log.Errorf("Unable to clean downloads dir %q after update: %v", paths.Downloads(), err)
+			u.log.Errorw("Unable to clean downloads after update", "error.message", err, "downloads.path", paths.Downloads())
 		}
+		u.log.Infow("Restarting after upgrade", "new_version", release.Version(), "prev_version", a.Version(),
```
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/upgrade.go:203]: consider debug, on other places as well. although upgrade is not that common operation a lot of new info messages here

**Comment by joshdover** [comment]
> The biggest issue I noticed was that we log action failures at the debug label by default, which includes the upgrade action. Upgrade action failures are infrequent and important enough that I switched to always logging them at error level.

Good find, I agree this alone will help a lot. We are seeing agents attempting upgrades and then restarting with no indication of why the upgrade failed. That said, this was happening even when debug logging was enabled which is why the other logs are important too so we can identify at which point there may be a process panic/crash.

File: internal/pkg/agent/application/upgrade/cleanup.go (hunk 1 above)
Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/cleanup.go:22]: I don't think we should do this around upgrades. Upgrade failures are some of the worst errors we can experience, and for failures that are hard to reproduce or leave the agent in a poor state you sometimes do not get the chance to go back and enable debug logs.

File: internal/pkg/agent/application/upgrade/rollback.go [hunk 5]
```diff
@@ -74,7 +76,9 @@ func Cleanup(currentHash string, removeMarker bool) error {
 	}
 
 	// remove symlink to avoid upgrade failures, ignore error
-	_ = os.Remove(prevSymlinkPath())
+	prevSymlink := prevSymlinkPath()
+	log.Infow("Removing previous symlink path", "file.path", prevSymlinkPath())
```
Comment (michel-laterman) [inline internal/pkg/agent/application/upgrade/rollback.go:80]: Should this be a debug line?

File: internal/pkg/agent/application/upgrade/step_download.go [hunk 6]
```diff
@@ -40,6 +40,10 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
 		}
 	}
 
+	u.log.Infow("Downloading upgrade artifact", "version", version,
```
Comment (michel-laterman) [inline internal/pkg/agent/application/upgrade/step_download.go:43]: I thought that the downloader emitted periodic lines, is this needed here as well?

**Review by michel-laterman — APPROVED** [review]
lgtm, however I would like to repeat Michal's comments that many of the entries should also probably be part of the debug logs

File: internal/pkg/agent/application/upgrade/step_download.go (hunk 6 above)
Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/step_download.go:43]: Yeah this one can be removed.

File: internal/pkg/agent/application/upgrade/rollback.go (hunk 5 above)
Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/rollback.go:80]: Yes, it isn't super valuable.

File: internal/pkg/agent/application/upgrade/upgrade.go (hunk 4 above)
Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/upgrade.go:203]: I downgraded a few of the less valuable messages to debug, we should catch most error cases regardless.

**Comment by cmacknz** [comment]
Merged, downgraded several messages to debug level as suggested.

**Pull request merged**